feat(rust): resolve forwarder project name in manager

twittner opened the pull request:
Since project routes are not guaranteed to be stable this commit moves the resolution of project references to routes when creating a remote forwarder to the node manager. The command-line layer merely maps a project name to a project ID but leaves everything else to the node manager.

On top of https://github.com/build-trust/ockam/pull/3497.

twittner committed:
feat(rust): resolve forwarder project name in manager

Since project routes are not guaranteed to be stable this commit moves
the resolution of project references to routes when creating a remote
forwarder to the node manager. The command-line layer merely maps a
project name to a project ID but leaves everything else to the node
manager.

twittner committed:
feat(rust): add `MultiAddr::matches`

Allows to quickly check if the address prefix matches a sequence of
codes.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/nodes/models/forwarder.rs:
```diff
@@ -20,35 +16,42 @@ pub struct CreateForwarder<'a> {
+            alias: alias.map(|s| s.into()),
+            at_rust_node: false,
+            cloud_addr: Some(cloud),
+            authorized: None,
```
Why `at_project` forwarder doesn't have authorized value?

twittner replied:
It uses the project authority.

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/forwarder/create.rs:
```diff
@@ -108,13 +90,11 @@ async fn rpc(ctx: Context, (opts, cmd): (CommandGlobalOpts, CreateCommand)) -> R
-            CredentialExchangeMode::Oneway,
-        );
+        let body = if Some(Project::CODE) == cmd.at.first().map(|p| p.code()) {
+            CreateForwarder::at_project(ma, Some(alias), cmd.cloud_opts.route())
```
Should we error if `cmd.authorized` was provided as well?

twittner replied:
Done in https://github.com/build-trust/ockam/pull/3528/commits/08d78aed923f83858d6996fcc98fa4a09f71a9a8

SanjoDeundiak commented on implementations/rust/ockam/ockam_command/src/forwarder/create.rs:
```diff
@@ -108,13 +90,11 @@ async fn rpc(ctx: Context, (opts, cmd): (CommandGlobalOpts, CreateCommand)) -> R
+        let body = if Some(Project::CODE) == cmd.at.first().map(|p| p.code()) {
+            CreateForwarder::at_project(ma, Some(alias), cmd.cloud_opts.route())
+        } else {
+            CreateForwarder::at_node(ma, Some(alias), at_rust_node, cmd.authorized)
```
Should we error if `cmd.cloud_opts` was provided as well?

twittner replied:
AFAIK they read from an environment variable and default to a standard address if not present. `CloudOpts` has no arguments to provide.

SanjoDeundiak commented on implementations/rust/ockam/ockam_api/src/util.rs:
```diff
@@ -127,29 +132,47 @@ pub fn multiaddr_to_addr(ma: &MultiAddr) -> Option<Address> {
+    Ok(ma)
+}
+
+pub fn is_secure_channel_addr(a: &MultiAddr) -> bool {
```
Does it check if secure channel will be created while using that MultiAddr, and doesn't detect if there is already existent secure channel in that MultiAddr? Would be nice to explain that in the comments (and maybe even in the name)

SanjoDeundiak replied:
Also wanted to clarify, is it possible that we want to check if next hop uses secure channel, e.g.
`/ipv4/127.0.0.1/tcp/1234/secure/api/ipv4/127.0.0.1/tcp4321/secure/api`, but for example this multiaddr also passes the check I guess? `/ipv4/127.0.0.1/tcp/1234/ipv4/127.0.0.1/tcp4321/secure/api`
I'm afraid of misusing that function

twittner replied:
To avoid confusion I have removed it in https://github.com/build-trust/ockam/pull/3528/commits/08d78aed923f83858d6996fcc98fa4a09f71a9a8. Reading the match inline seems to be more obvious to readers.

twittner committed:
fix(rust): review feedback

- Error if --authorized is used with project addresses.
- Remove `is_secure_channel_addr` function.
- Add offset to `MultiAddr::match`.

twittner commented:
@SanjoDeundiak: Could you review again please?
